aea/helpers/pypi_checker.py

```diff
@@ -0,0 +1,297 @@
+    - "<=<some-version-number>"
+    - ">=<some-version-number>"
+
+    The equality might be optional.
+
+    The pseudo-code is the following:
+    - is the version number of the lower-than constraint smaller than the one of
+      the greater-than constraint? E.g. we are in cases like "<=1.0" and ">1.1".
+      -  If yes, then the constraints are unsatisfiable.
+      -  Otherwise, they are satisfiable.
+    - are the version numbers the same? E.g. "<=1.0,>=1.0"
+      - If yes, if at least one of them is a strict comparison, then return False, otherwise True.
+    - otherwise, return True.
+
+    :param lowest_less_than: the less-than constraint.
+    :param greatest_greater_than: the greater than constraint.
+    :return: False if we are sure the two constraints are not satisfiable, True if we don't know.
+    """
+    version_less_than = Version(lowest_less_than.version)
+    version_greater_than = Version(greatest_greater_than.version)
+    if version_less_than < version_greater_than:
+        return False
+    elif version_greater_than == version_less_than:
+        # check if one of them has NOT the equality
+        one_of_them_is_a_strict_comparison = (
+            greatest_greater_than.operator == ">"
+        ) or (lowest_less_than.operator == "<")
+        return not one_of_them_is_a_strict_comparison
+    else:
+        return True
+
+
+class PyPIChecker:
+    """
+    Find out if PyPI version specifiers are unsatisfiable.
+
+    Its usage is incremental: you keep an instance of the
+    checker and you populate with new constraints.
+    """
+
+    def __init__(self):
+        """Initialize the checker."""
+        # mapping from package name to specifier set.
+        self._pypi_dependencies = {}  # type: Dict[str, SpecifierSet]
+
+    def add(self, package_name: str, specifier_set: SpecifierSet):
+        """
+        Add a specifier set for a package in the checker.
+
+        :param package_name: the package name
+        :param specifier_set: the set of specifiers.
+        :return: None
+        :raises ValueError: if the set of specifier plus the one already present
+            is surely unsatisfiable.
+        """
+        specifier = self._merge_with_existing_specifier_set_if_any(
+            package_name, specifier_set
+        )
+
+        try:
+            result = _is_satisfiable(specifier)
+        except Exception:
+            # we ignore possible errors. There might be some corner cases that make
+            # our checks to raise exception, e.g. requirements like git+https://...
+            result = True
+
+        if result is False:
+            raise ValueError(
+                "Specifier set {} for package {} not satisfiable.".format(
+                    specifier_set, package_name
+                )
+            )
+        self._pypi_dependencies[package_name] = specifier_set
+
+    def _add_or_merge_with_existing_specifier_set(
```

Comment:
What is the use case for this method compared to `_merge_with_existing_specifier_set_if_any`?